---
title: "Preprocessing Report"
author: "Leon Wang"
date: "6 Jan, 2026"
format: html
toc: true
---

```{r}
#| label: load-packages
#| include: false

# library(shiny)
library(tidyverse)
library(data.table)
```

## Introduction

Thank you for running this Nextflow pipeline for preprocessing Oxford Nanopore antibody sequencing data.

## 1. Aligned reads

The table below provides a summary after mapping and alignment of target antibody sequences to the reference genome provided, using the [*minimap2 package*](https://lh3.github.io/minimap2/) (v2.30) and the [*samtools package*](https://www.htslib.org) (v1.22.1).

```{r}
# Read in data
#| include: false

# Get list of files
files <- list.files(
    path = "./results/1. aligned reads/stats",
    pattern = "\\.tsv$",
    full.names = TRUE
)

# Read in files
data_list <- lapply(files, fread, na.strings = c("N/A"))
# Add in barcode from file name as id col
names(data_list) <- str_replace(basename(files), "_alignment_stats.tsv", "")

# Combine datasets together and rename columns
combined_data <- rbindlist(data_list, idcol = "barcode", fill = TRUE)
combined_data <- rename(combined_data, "QC_pass" = V1, "QC_fail" = V2, "category" = V3)

# Rename and extract required rows
rows_required <- list("total", "mapped", "mapped %", "primary", "secondary", "supplementary")
combined_data <- combined_data %>%
    mutate(category = ifelse(category == "total (QC-passed reads + QC-failed reads)",
        "total", category
    )) %>%
    mutate(across(
        c(QC_pass, QC_fail),
        ~ ifelse(str_detect(.x, "%"),
            as.numeric(str_replace(.x, "%", "")) / 100,
            as.numeric(.x)
        )
    )) %>%
    filter(category %in% rows_required)
```

```{r}
combined_data %>%
    distinct(barcode) %>%
    summarise(n = n())
```

```{r}
#| label: total-reads

ggplot(data = filter(combined_data, category == "total"), aes(y = barcode, x = QC_pass)) +
    geom_col()
```

```{r}
#| label: mapped-perc

ggplot(data = filter(combined_data, category == "mapped %"), aes(x = QC_pass)) +
    geom_density()
```

## 2. Extracted reads

The followin provides detail after extracting the heavy and light chain sequences using matchbox [*samtools package*](https://jakob-schuster.github.io/matchbox-docs/).

```{r}
# Read in data
#| include: false

# Get list of files
files <- list.files(
    path = "./results/1. aligned reads/stats",
    pattern = "\\.tsv$",
    full.names = TRUE
)

# Read in files
data_list <- lapply(files, fread, na.strings = c("N/A"))
# Add in barcode from file name as id col
names(data_list) <- str_replace(basename(files), "_alignment_stats.tsv", "")

```

## 3. Annotated reads

The table below provides a summary after mapping and alignment of target antibody sequences to the reference genome provided, using the [*minimap2 package*](https://lh3.github.io/minimap2/) (v2.30) and the [*samtools package*](https://www.htslib.org) (v1.22.1).